### DISCUSSION TITLE
Debugging when developing a custom sf plugin and it fails to compile

### DISCUSSION THREAD

**Discussion by tomcarman** [forcedotcom/cli#2917] [Q&A] [answered]
When building a custom sf plugin, sometimes the plugin fails to even compile - eg.

Running:
`bin/dev.js mycommand run`

Gives:

```
Error: command mycommand:run not found
    at Config.runCommand (some-project/node_modules/@oclif/core/lib/config/config.js:403:19)
    at async run (some-project/node_modules/@oclif/core/lib/main.js:92:16)
    at async main (file://some-project/bin/dev.js:5:3)
    at async file://some-project/bin/dev.js:8:1

```

This is normally down to things like wonky imports after a refactor - but I'm always at a loss at how to debug. Eslint is generally reporting everything is fine, and without being able to even run the plugin, I can't use the normal debugging options.

Possibly a stupid question (I'm fairly new to JS/TS) - but whats the right thing to do when this happens, other than spending a long time trying random things until I've worked out whats wrong?

**Comment by WillieRuemmele** [comment]
Hey @tomcarman  - 

Are you running `yarn build` or something similar before? You need to have your source code compiled into the `lib` directory usually. You might try deleting that and recompiling. Make sure to pay attention to those command outputs to make sure everything built correctly - otherwise OCLIF will fail to find your command, and you'll see that error

**Comment by tomcarman** [comment, marked as answer]
Hi @WillieRuemmele, thanks for the reply.

I tried deleting the `lib` directory and recompiling with `yarn build`. It builds with no warnings / errors, but still get the same error when trying to run a command.

For additional context, its a plugin generated with the `sf dev` plugin. And I am trying to execute my command locally using the provided `bin/dev.js` eg. this https://github.com/salesforcecli/plugin-template-sf-external/blob/main/bin/dev.js

This works great most of the time, but occasionally I manage to introduce an issue when developing thats not picked up by eslint, and causes the command to fail to properly compile, giving the error in the initial post.

I've had this a few times now, and normally just have to end up reverting changes one-by-one until I can work out where its falling over. But it would great if I could get some better debug information to make this easier. Any other ideas?

**Edit**

I've actually just tried something that I think is pointing me in the right direction, I just tried executing the command .js file directly eg:

`node lib/commands/somecommand.js` 

...and I can now see the error causing the issue - I introduced some circular dependencies when refactoring.

**Reply by WillieRuemmele** [reply]
Interesting - I don't think I've personally run across a circular dependency issue while trying to run a command locally. Sounds like you learned a good trick, I'll try and remember that for the next time I see this error. @k-capehart's suggestions are also very relevant, and common to run while developing a plugin.

**Comment by k-capehart** [comment]
Looks like you figured this out, but I've found that sometimes a `yarn clean-all` followed by a `yarn install` and `yarn build` will get things back up and running for me.
